src/mame/misc/astrcorp.cpp

```diff
@@ -1,58 +1,168 @@
-- winbingo and clones, zulu: these should be at the same state of the two above, but don't show anything on screen;
-- magibomba, westvent: need a redump of one of the program ROMs;
+- MCU on showhand/showhanc PCB, what does it do? Games run fine and there appears to be no interaction...
+- Protection in magibomb sets with simpler encryption (writes to $a0101x, $a11100, etc.). Some sets appear to work fine regardless, others freeze.
+- Fix mirror ROM checksum / ROM overlay without code patches (in games with heavier encryption).
+- Find source of level 2 interrupt (sprite DMA end?).
+- magibomba, westvent: need a redump of one of the program ROMs.
+- magibombd, hacher: need a redump of the sprite ROMs.
+- astoneag, dinodino, magibombd: exiting from test menu goes haywire (requires a soft-reset with F3).
 
 *************************************************************************************************************/
 
 #include "emu.h"
+
 #include "cpu/m68000/m68000.h"
 #include "machine/eepromser.h"
 #include "machine/nvram.h"
 #include "machine/ticket.h"
 #include "machine/timer.h"
 #include "sound/okim6295.h"
 #include "video/ramdac.h"
+
 #include "emupal.h"
 #include "screen.h"
 #include "speaker.h"
 
+#include "astoneag.lh"
+#include "dinodino.lh"
+#include "magibomb.lh"
+#include "magibombb.lh"
+#include "showhanc.lh"
+#include "showhand.lh"
+#include "skilldrp.lh"
+#include "winbingo.lh"
+#include "zoo.lh"
+
+/***************************************************************************
+
+    CPU Code (Protection)
+
+***************************************************************************/
+
+class astro_cpucode_device : public device_t
+{
+public:
+	astro_cpucode_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock);
+
+	// read handlers
+	DECLARE_READ_LINE_MEMBER(do_read);    // DO
+
+	// write handlers
+	DECLARE_WRITE_LINE_MEMBER(cs_write);  // CS signal (active high)
+	DECLARE_WRITE_LINE_MEMBER(clk_write); // CLK signal (active high)
+
+protected:
+	virtual void device_start() override;
+	virtual void device_reset() override;
+
+private:
+	required_memory_region m_region_key;
+
+	u8 m_cs_state = CLEAR_LINE;
+	attotime m_last_cs_rising_edge_time = attotime::zero;
+	u8 m_clk_state = CLEAR_LINE;
+	u16 m_shift_register = 0;
+};
+
+DEFINE_DEVICE_TYPE(ASTRO_CPUCODE, astro_cpucode_device, "astro_cpucode", "Astro CPU Code")
+
+astro_cpucode_device::astro_cpucode_device(const machine_config &mconfig, const char *tag, device_t *owner, u32 clock)
+	: device_t(mconfig, ASTRO_CPUCODE, tag, owner, clock)
+	, m_region_key(*this, DEVICE_SELF)
+{ }
+
+void astro_cpucode_device::device_start()
+{
+	save_item(NAME(m_cs_state));
+	save_item(NAME(m_last_cs_rising_edge_time));
+	save_item(NAME(m_clk_state));
+	save_item(NAME(m_shift_register));
+}
+
+void astro_cpucode_device::device_reset()
+{
+	m_cs_state = m_clk_state = CLEAR_LINE;
+	m_shift_register = 0;
+}
```

Comment:
You shouldn’t change input states in `device_reset` – if the thing driving the inputs changes the state of its outputs on reset, the new state will get pushed through.  Doing this can make things get out-of-sync.

Reply:
I see. Removed the input states changes.